docs/eBpfExtensions.md

```diff
@@ -1,42 +1,207 @@
 
+## 1.3 NPI Contracts for eBPF Extensions
+eBPF Extensions need to implement *provider modules* for two types of NPIs. They are the **Program Information NPI** provider and the **Hook NPI** provider.
+
+### 1.3.1 eBPF Program Information NPI Provider
 The eBPF program information NPI contract is used to provide information about an eBPF program type. Program types
 are defined as the ABI contracts that eBPF programs are written to conform to.
 This information is consumed by the eBPF verifier to ensure that any eBPF programs of a given type are safe to load
 and execute. The ABI contract includes both a description of the &quot;context&quot; parameter passed to the eBPF
-program as well as the list of specific helper functions that are available to such eBPF programs. eBPF extensions
-should not unload until the last NMR client detaches, signifying that no eBPF programs are using helper functions
-it provides.
+program as well as the list of specific helper functions that are available to such eBPF programs.
+
+eBPF extensions must implement a different Program Information provider module for each program type for which it provides
+implementation of hooks or helper functions specific to that program type.
+
+### 1.3.2 eBPF Hook NPI Provider
+The Hook NPI contracts are used by extension to invoke eBPF programs when OS events occur. These are also known as *Attach NPI* and is used
+interchangeably along with *Hook NPI* in the code and documentations in this project. A given eBPF hook may have several attach types. The eBPF extension must register a separate Hook or Attach NPI provider module for each attach type it supports for an eBPF hook.
+
+## 2 Authoring an eBPF Extension
+The steps for authoring an eBPF extension are:
+1. Registering the NPI providers.
+2. Authoring Helper Functions.
+3. Invoking eBPF programs from hook(s).
+4. Registering program and attach types.
+
+The following sections describe these steps in detail.
 
-Currently the information is exposed as an opaque blob of
-[MS-RPCE](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rpce/290c38b1-92fe-4229-91e6-4fc376610c15)
-serialized data with the schema defined by
-[ebpf_program_types](https://github.com/microsoft/ebpf-for-windows/blob/master/libs/platform/ebpf_program_types.idl).
-Due to current limitations on Windows, it's not possible to serialize the contract from kernel mode. Proposals are
-underway to switch the serialization to use [Google FlatBuffers](https://google.github.io/flatbuffers/) as that is
-more easily serialized from kernel mode.
+### 2.1 Program Information NPI Provider Registration
+When registering itself to the NMR, the Program Information NPI Provider should have the [`NPI_REGISTRATION_INSTANCE`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/netioddk/ns-netioddk-_npi_registration_instance) initialized as follows:
```

Comment:
nit: This line has "Program Information NPI", but line 27 has "program information NPI".  Pick one and be consistent throughout the doc.

Reply:
Fixed. Changed to "Program Information NPI" and "Hook NPI" all over.